File: Frontend/types/proposal.ts
import type { BudgetType } from "./job"

export type ProposalStatus = "new" | "shortlisted" | "hired" | "declined"

export interface Proposal {
  id: string
  job_id: string
  freelancer_id: string
  cover_letter: string
  bid: number
  budget_type: BudgetType
  delivery_time: string
  status: ProposalStatus
  client_message?: string
  submitted_at: string
  viewed_at?: string | null
  responded_at?: string | null
  created_at: string
  updated_at: string

  // Joined — job / client
  job_title: string
  client_name: string
  category: string

  // Joined — freelancer
  freelancer_name: string
  freelancer_role: string
  freelancer_tier: string
  freelancer_rating: number
  freelancer_reviews: number
  freelancer_verified: boolean
  freelancer_avatar?: string | null

  // Computed
  color?: string
}

export interface ProposalCreateInput {
  job_id: string
  cover_letter: string
  bid: number
  budget_type: BudgetType
  delivery_time: string
}

export interface UpdateProposalStatusInput {
  status: ProposalStatus
}

File: Backend/internal/entity/proposal.go
package entity

import (
	"time"

	"github.com/google/uuid"
)

type Proposal struct {
	ID                uuid.UUID  `db:"id" json:"id"`
	JobID             uuid.UUID  `db:"job_id" json:"job_id"`
	FreelancerID      uuid.UUID  `db:"freelancer_id" json:"freelancer_id"`
	CoverLetter       string     `db:"cover_letter" json:"cover_letter"`
	BidAmount         float64    `db:"bid_amount" json:"bid"`
	BudgetType        string     `db:"budget_type" json:"budget_type"`
	DeliveryTimeLabel string     `db:"delivery_time_label" json:"delivery_time"`
	Status            string     `db:"status" json:"status"`
	ClientMessage     string     `db:"client_message" json:"client_message"`
	SubmittedAt       time.Time  `db:"submitted_at" json:"submitted_at"`
	ViewedAt          *time.Time `db:"viewed_at" json:"viewed_at,omitempty"`
	RespondedAt       *time.Time `db:"responded_at" json:"responded_at,omitempty"`
	CreatedAt         time.Time  `db:"created_at" json:"created_at"`
	UpdatedAt         time.Time  `db:"updated_at" json:"updated_at"`

	// Joined — job / client
	JobTitle   string `db:"job_title" json:"job_title"`
	ClientName string `db:"client_name" json:"client_name"`
	Category   string `db:"category" json:"category"`

	// Joined — freelancer
	FreelancerName     string   `db:"freelancer_name" json:"freelancer_name"`
	FreelancerRole     string   `db:"freelancer_role" json:"freelancer_role"`
	FreelancerTier     string   `db:"freelancer_tier" json:"freelancer_tier"`
	FreelancerRating   float64  `db:"freelancer_rating" json:"freelancer_rating"`
	FreelancerReviews  int      `db:"freelancer_reviews" json:"freelancer_reviews"`
	FreelancerVerified bool     `db:"freelancer_verified" json:"freelancer_verified"`
	FreelancerAvatar   *string  `db:"freelancer_avatar" json:"freelancer_avatar,omitempty"`

	// Computed
	Color string `db:"-" json:"color"`
}

type ProposalCreateInput struct {
	JobID             string  `json:"job_id"`
	CoverLetter       string  `json:"cover_letter"`
	BidAmount         float64 `json:"bid"`
	BudgetType        string  `json:"budget_type"`
	DeliveryTimeLabel string  `json:"delivery_time"`
}
